### ISSUE TITLE
Panic on recovery synced shard

### ISSUE THREAD

**Issue by generall** [qdrant/qdrant#2184]
If the service is killed during the shard trasnfer, the next load of the server may end up with panic:

<details>
  <summary>stack</summary>
  
 ```
  Recovering collection benchmark [00:00:09] ███████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 203/334 (eta:6s)
  [2023-07-01T08:51:40.188Z ERROR qdrant::startup] Panic backtrace: 
       0: qdrant::startup::setup_panic_hook::{{closure}}
                 at ./src/startup.rs:38:25
       1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/boxed.rs:1987:9
       2: std::panicking::rust_panic_with_hook
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:695:13
       3: std::panicking::begin_panic_handler::{{closure}}
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:582:13
       4: std::sys_common::backtrace::__rust_end_short_backtrace
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:150:18
       5: rust_begin_unwind
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
       6: core::panicking::panic_fmt
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
       7: core::slice::index::slice_end_index_len_fail_rt
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:78:5
       8: core::slice::index::slice_end_index_len_fail
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:70:9
       9: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:411:13
      10: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:19:9
      11: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/vec/mod.rs:2691:9
      12: segment::vector_storage::chunked_vectors::ChunkedVectors<T>::get
                 at ./lib/segment/src/vector_storage/chunked_vectors.rs:52:10
      13: <segment::vector_storage::simple_vector_storage::SimpleVectorStorage as segment::vector_storage::vector_storage_base::VectorStorage>::get_vector
                 at ./lib/segment/src/vector_storage/simple_vector_storage.rs:147:9
      14: <segment::vector_storage::vector_storage_base::VectorStorageEnum as segment::vector_storage::vector_storage_base::VectorStorage>::get_vector
                 at ./lib/segment/src/vector_storage/vector_storage_base.rs:164:45
      15: segment::segment::Segment::vector_by_offset
                 at ./lib/segment/src/segment.rs:382:17
      16: <segment::segment::Segment as segment::entry::entry_point::SegmentEntry>::vector
                 at ./lib/segment/src/segment.rs:964:26
      17: <segment::segment::Segment as segment::entry::entry_point::SegmentEntry>::all_vectors
                 at ./lib/segment/src/segment.rs:971:32
      18: collection::collection_manager::segments_updater::sync_points::{{closure}}
                 at ./lib/collection/src/collection_manager/segments_updater.rs:298:27
      19: collection::collection_manager::holders::segment_holder::SegmentHolder::read_points
                 at ./lib/collection/src/collection_manager/holders/segment_holder.rs:387:29
      20: collection::collection_manager::segments_updater::sync_points
                 at ./lib/collection/src/collection_manager/segments_updater.rs:297:24
      21: collection::collection_manager::segments_updater::process_point_operation
                 at ./lib/collection/src/collection_manager/segments_updater.rs:433:43
      22: collection::collection_manager::collection_updater::CollectionUpdater::update
                 at ./lib/collection/src/collection_manager/collection_updater.rs:55:17
      23: collection::shards::local_shard::LocalShard::load_from_wal
                 at ./lib/collection/src/shards/local_shard.rs:436:20
      24: collection::shards::local_shard::LocalShard::load::{{closure}}
                 at ./lib/collection/src/shards/local_shard.rs:241:9
      25: collection::shards::replica_set::ShardReplicaSet::load::{{closure}}
                 at ./lib/collection/src/shards/replica_set.rs:496:17
      26: collection::shards::shard_holder::ShardHolder::load_shards::{{closure}}
                 at ./lib/collection/src/shards/shard_holder.rs:210:17
      27: collection::collection::Collection::load::{{closure}}
                 at ./lib/collection/src/collection.rs:262:13
      28: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/park.rs:283:63
      29: tokio::runtime::coop::with_budget
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/coop.rs:107:5
      30: tokio::runtime::coop::budget
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/coop.rs:73:5
      31: tokio::runtime::park::CachedParkThread::block_on
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/park.rs:283:31
      32: tokio::runtime::context::blocking::BlockingRegionGuard::block_on
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/context/blocking.rs:66:9
      33: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/scheduler/multi_thread/mod.rs:87:13
      34: tokio::runtime::context::runtime::enter_runtime
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/context/runtime.rs:65:16
      35: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
      36: tokio::runtime::runtime::Runtime::block_on
                 at /home/generall/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.0/src/runtime/runtime.rs:313:45
      37: storage::content_manager::toc::TableOfContent::new
                 at ./lib/storage/src/content_manager/toc.rs:142:30
      38: qdrant::main
                 at ./src/main.rs:213:15
      39: core::ops::function::FnOnce::call_once
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:250:5
      40: std::sys_common::backtrace::__rust_begin_short_backtrace
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:134:18
      41: std::rt::lang_start::{{closure}}
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/rt.rs:166:18
      42: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:287:13
      43: std::panicking::try::do_call
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:485:40
      44: std::panicking::try
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:449:19
      45: std::panic::catch_unwind
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panic.rs:140:14
      46: std::rt::lang_start_internal::{{closure}}
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/rt.rs:148:48
      47: std::panicking::try::do_call
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:485:40
      48: std::panicking::try
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:449:19
      49: std::panic::catch_unwind
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panic.rs:140:14
      50: std::rt::lang_start_internal
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/rt.rs:148:20
      51: std::rt::lang_start
                 at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/rt.rs:165:17
      52: main
      53: <unknown>
      54: __libc_start_main
      55: _start
    
[2023-07-01T08:51:40.188Z ERROR qdrant::startup] Panic occurred in file lib/segment/src/vector_storage/chunked_vectors.rs at line 52: range end index 6346752 out of range for slice of length 6338560

```
</details>

## Current Behavior

Panics on recovery

## Steps to Reproduce

1. Run in cluster mode
2. Run shards transfer from one node to another, having the receiver empty before transfer
3. Kill the recover

## Expected Behavior

no panic

## Possible Solution

It looks like, that `sync` API tries to read the point before inserting it. But the inconsistent state after the kill might contain point ids without associated vectors (In this case it assumes recovery from WAL). That results in panic.

## Possible Implementation

Options:
- Allow reading non-existing vectors instead of panic (return None instead)
- Use special checks read in `sync` API to prevent panic

**Comment by generall** [comment]
Snapshot to reproduce: https://storage.googleapis.com/qdrant-debug/storage-panic.tar.gz

**Comment by generall** [comment]
fixed in 1.3.1
